lib/OpenQA/WebAPI/Controller/API/V1/JobTemplate.pm

```diff
@@ -119,6 +121,136 @@ sub list {
+
+    # Note: Using the schema filename; slurp'ed text isn't detected as YAML
+    my $schema_yaml = $self->app->home->child('public', 'schema', 'JobTemplate.yaml')->to_string;
+    my $validator   = JSON::Validator->new;
+    my $schema;
+    my @errors;
+    if ($validate_schema) {
+        # Validate the schema: catches errors in type names and definitions
+        $validator = $validator->load_and_validate_schema($schema_yaml);
+        $schema    = $validator->schema;
+    }
+    else {
+        $schema = $validator->schema($schema_yaml);
+    }
+    if ($schema) {
+        # Note: Don't pass $schema here, that won't work
+        push @errors, $validator->validate($yaml);
+    }
+    else {
+        push @errors, "Failed to load schema";
+    }
+    \@errors;
+}
+
+=over 4
+
+=item schedules()
+
+Serializes the given job group with relevant test suites by architecture and products (mediums), or all available
+groups defined in the system if no group id is specified.
+Common defaults for prio and machine are represented in the defaults key.
+
+Returns a YAML template representing the job groups(s).
+
+=back
+
+=cut
+
+sub schedules {
+    my $self = shift;
+
+    my %yaml;
+    my $groups = $self->db->resultset("JobGroups")
+      ->search($self->param('id') ? {id => $self->param('id')} : undef, {select => [qw(id name)]});
+    while (my $group = $groups->next) {
+        my %group;
+        my $templates
+          = $self->db->resultset("JobTemplates")->search({group_id => $group->id}, {order_by => 'me.test_suite_id'});
+
+        # Always set the hash of test suites to account for empty groups
+        $group{architectures} = {};
+        $group{products}      = {};
+
+        my %machines;
+        my %prios;
+        # Extract products and tests per architecture
+        while (my $template = $templates->next) {
+            $group{products}{$template->product->name} = {
+                distri  => $template->product->distri,
+                flavor  => $template->product->flavor,
+                version => $template->product->version
+            };
+            my %test_suite;
+            $test_suite{machine} = $template->machine->name;
+            $machines{$template->product->arch}{$template->machine->name}++;
+            if ($template->prio) {
+                $test_suite{prio} = $template->prio;
+                $prios{$template->product->arch}{$template->prio}++;
+            }
+            my $test_suites = $group{architectures}{$template->product->arch}{$template->product->name};
+            push @$test_suites, {$template->test_suite->name => \%test_suite};
+            $group{architectures}{$template->product->arch}{$template->product->name} = $test_suites;
+        }
+
+        # Split off defaults
+        foreach my $arch (keys %{$group{architectures}}) {
+            # Ensure prios hash is defined for this architecture
+            $prios{$arch}{0} = 0;
+            my %_prios       = %{$prios{$arch}};
```

Comment:
The underscore looks a bit out of place. Maybe name it `%arch_prios`?

Reply:
I removed the helper variables